hybrasyl/Objects/User.cs

```diff
@@ -1260,35 +1260,174 @@ public void SendVisibleGold(Gold gold)
+                {
+                    Length = (uint)castable.Cooldown,
+                    Pane = 1,
+                    Slot = slot
+                }.Packet()); ;
+            }
         }
 
         internal void UseSpell(byte slot, uint target = 0)
         {
             var castable = SpellBook[slot];
-            Creature targetCreature = Map.EntityTree.OfType<Monster>().SingleOrDefault(x => x.Id == target) ?? null;
-            Direction playerFacing = this.Direction;
-            if (targetCreature is Merchant) return;
-            if (targetCreature != null) Attack(castable, targetCreature);
-            else Attack(castable);
+            Creature targetCreature = Map.EntityTree.OfType<Creature>().SingleOrDefault(x => x.Id == target) ?? null;
+            if (canCast(castable))
+            {
+                if(castable.Effects.Damage != null)
+                {
+                    List<Creature> targets = new List<Creature>();
+                    bool spellCast = false;
+                    if (targetCreature != null)
+                    {
+                        targets.Add(targetCreature);
+                    }
+                    else
+                    {
+                        targets = PossibleTargets(castable);
+                    }
 
-            Client.Enqueue(new ServerPacketStructures.Cooldown()
+                    foreach(var tar in targets)
+                    {
+                        bool canAttack = !(tar is Merchant || (tar is User && (!(tar as User).Status.HasFlag(PlayerCondition.Pvp))));
+                        if (canAttack)
+                        {
+                            Attack(castable, tar);
+                            spellCast = true;
+                        }
+                    }
+                    if(spellCast) CastingCost(castable);
+                }
+                else if(castable.Effects.Heal != null)
+                {
+                    List<Creature> targets = new List<Creature>();
+                    bool spellCast = false;
+                    if (targetCreature != null)
+                    {
+                        targets.Add(targetCreature);
+                    }
+                    else
+                    {
+                        targets = PossibleTargets(castable);
+                    }
+
+                    foreach (var tar in targets)
+                    {
+                        bool canHeal = !(tar is Monster);
+                        if (canHeal)
+                        {
+                            HealTarget(castable, tar);
+                            spellCast = true;
+                        }
+                    }
+                    if(spellCast) CastingCost(castable);
+                }
+                else
+                {
+                    this.SendSystemMessage("Having no objective in mind, the spell dissipates.");
+                }
+            }
+        }
+
+        public bool canCast(Castable castable)

```

Comment:
this method should just be a check. it should not affect actual values.

That should be handled within the appropriate attack/heal methods.

Reply:
I don't see where canCast is affecting actual values. Can you please confirm?